src/checkers/inference/InferenceQualifierHierarchy.java

```diff
@@ -224,9 +226,15 @@ public boolean isSubtype(final AnnotationMirror subtype, final AnnotationMirror
 
         final Slot subSlot   = slotMgr.getSlot(subtype);
         final Slot superSlot = slotMgr.getSlot(supertype);
-        constraintMgr.addSubtypeConstraint(subSlot, superSlot);
 
-        return true;
+        Constraint constraint = constraintMgr.createSubtypeConstraint(subSlot, superSlot);
+
+        if (constraint instanceof AlwaysFalseConstraint) {
```

Comment:
This is not conceptually correct even though in implementation it may work. We need to undertake the true meaning of inferring programs. Also, ConstraintManager#add() should never be public and called by programmer to add lingering constraint into ConstraintManager. We can discuss about this in tomorrow’s meeting.

Reply:
The only difference to what you had is that this code should also check for `AlwaysTrueContraint` and directly return true.
At the moment the code here adds the `AlwaysTrueConstraint` to the constraint set.

Won't we need a public `add` method to add implication constraints as well?
I agree that the current `add` method that takes a class literal shouldn't be public.

Reply:
> The only difference to what you had is that this code should also check for AlwaysTrueContraint and directly return true.
At the moment the code here adds the AlwaysTrueConstraint to the constraint set.

From my point of view, this is not the essential difference. For me, we shouldn't try to create any constraint at all when two slots are `ConstantSlot`.

Why do we generate constraints in inference? This is because in unannotated programs, there should be a place holder to represent the missing annotation, and we add constraints over those "place holders" and admit that the related code typechecks and continue typechecking the program(or inferring).
But if two slots are both Constant, that means we already have all the annotations needed. That means: 1) we don't need to use any place holder - VariableSlot to represent an unknown value. 2) we don't need to generate Constraints, as there is nothing to be solved/unknown.

> Won't we need a public add method to add implication constraints as well?

No, `ConstraintManager` should have another API - `void addImplicationConstraint(Constraint... assumptions, Constraint conclusion)` to add `ImplicationConstraint`.
The only use case of `createConstraint()` is to create component `Constraint`s that make up of `ImplicationConstraint`. But now, in order to test the case that has two `ConstantSlot`s, we call `createConstraint()`, which is weird.

I think no "add()" method in `ConstraintManager` should be public, as it causes two ways of adding `Constraint`s - 1) Call `createConstraint()`, then call `add()` 2) Call `addConstraint()`, which is not nice as we'll have non-orthogonal APIs.

Reply:
> in order to test the case that has two ConstantSlots, we call createConstraint() 

Isn't this just `subSlot.isConstant()` ?

The normalized constraint is created to test for true/false.

I pushed an update to address the class literal problem, and it delays normalization until the private `ConstraintManager.add()` is called.

> But if two slots are both Constant, that means we already have all the annotations needed. That means: 1) we don't need to use any place holder - VariableSlot to represent an unknown value. 2) we don't need to generate Constraints, as there is nothing to be solved/unknown.

In the way this PR is designed, all constant-constant pairs get normalized to either true or false, and then both of those do not get inserted into the constraint set. So ultimately it achieves what you want.

To me the logic that *because* we have two constant pairs then we don't create a constraint is counter-intuitive. A constraint represents a type rule, or a transfer function (VPA/Arithmetic) that is part of a type rule. In type checking mode, we don't skip any rules and nor should we "not generate" constraints in inference. To me, constraint should always be uniformly and consistently generated whenever we type check. The fact that some specific pair of constant-constant values evaluates to always true or false is a special case that can be optimized through normalization.

Architecturally, we concentrate all the check logic between constant-constant pairs into one place - the constraint classes themselves since each constraint should know and fully define how it handles constant-constant pairs. It would make no sense to have constant-constant check code appearing all over the code base.

Lastly, to me the constraint manager should only be responsible for managing the constraint set and guaranteeing that the set doesn't have redundant (always true) or impossible (always false) constraints.

Reply:
> A constraint represents a type rule

Not quite agree with this opinion. Type rule is type rule, constraint is constraint. We have the same type rule in both typechecking and inference. It doesn't depend on constraints, nor does it have equal relation with constraints. Constraints are generated so that the annotation to be inferred obeys the type rule. So they aren't the same.

Once you've generated constraints, that means you have "admitted" the related code obeys the type rules already, and it's safe to continue the same typechecking/inference on the next lines. But, if the slots are both constants, and subtype relationship doesn't hold, you can revert the generated constraint and abort it. Then why generating/creating constraints? Just to test whether the two consts are really subtypes or not?

> It would make no sense to have constant-constant check code appearing all over the code base

We could have a separate class that checks the case for two-const cases to remove duplicate code(although I think there aren't except in `isSubtype()` and `leastUpperBound()`).
Even in this PR, each `XXXConstraint` class also has duplicate `if( xx instance of ConstantSlot && yy instanceof ConstantSLot)`. So we didn't completely opt out duplicate.
Also, it seems like we solved the `isSubtype()` problem by refactoring `ConstraintManager`, but how are you going to remove this line using the same logic?
https://github.com/opprop/checker-framework-inference/blob/358e623daf589995df42fbfd3b6b7a1e592be33d/src/checkers/inference/InferenceQualifierHierarchy.java#L266
Do we also tweak `slotMgr.createLubVariableSlot(var1, var2)` to return `ConstantSlot` if both `var1` and `var2` are `Constant`?

Reply:
To me a slot represents an annotation mirror (either known or to be given by solver), and a constraint represents a type rule.

Assuming your comments about Line 266 is after we create LUBconstraint, LUB varslot, etc, then the code in the entire method should just be create a LUBvarslot and then a LUBconstraint between the 3 slots, and return a reference to the annotation mirror in the LUBvarslot. The constant-constant case would be computed within LUBconstraint, updating the annotation mirror in LUBvarslot appropriately.